Add augmentation mri transforms

georgeyiasemis opened the pull request:
This PR:
* Adds `RescaleKspace`, `PadKspace` trasnforms (these project the data to image domain, perform augmentation, and then transform back to k-space)
* Adds `CompressCoilModule` transform which compresses the number of k-space coils using SVD
* Adds tests for new transforms
* Updates workflows for python v3.9 and v3.10

jonasteuwen commented on contributing.rst:
```diff
@@ -110,7 +110,7 @@ Before you submit a pull request, check that it meets these guidelines:
 #. If the pull request adds functionality, the docs should be updated. Put
    your new functionality into a function with a docstring, and add the
    feature to the list in README.rst.
-#. The pull request should work for Python 3.8 and 3.9 and for PyPy. Check Github actions and see that all tests pass.
+#. The pull request should work for Python 3.9 and for PyPy. Check Github actions and see that all tests pass.
 
```
3.10

jonasteuwen commented on direct/data/datasets_config.py:
```diff
@@ -37,9 +40,17 @@ class SensitivityMapEstimationTransformConfig(BaseConfig):
+
 @dataclass
 class RandomAugmentationTransformsConfig(BaseConfig):
-    random_rotation_degrees: Tuple[int, ...] = (-90, 90)
+    random_rotation_degrees: tuple[int, ...] = (-90, 90)
     random_rotation_probability: float = 0.0
```
Is this not tuple[int, int]?

georgeyiasemis replied:
No it can be of any length, eg. (90, ) or (-90, -180, 90), etc

jonasteuwen commented on direct/data/transforms.py:
```diff
@@ -34,7 +42,7 @@ def to_tensor(data: np.ndarray) -> torch.Tensor:
     return torch.from_numpy(data)
 
 
-def verify_fft_dtype_possible(data: torch.Tensor, dims: Tuple[int, ...]) -> bool:
+def verify_fft_dtype_possible(data: torch.Tensor, dims: tuple[int, ...]) -> bool:
     """fft and ifft can only be performed on GPU in float16 if the shapes are powers of 2. This function verifies if
```
I think you can make this tuple[int, int] | tuple[int, int, int]? You can define that type somewhere else and use it here

georgeyiasemis replied:
Done. I didnt define it as a new type only using it there for now.

jonasteuwen commented on direct/data/transforms.py:
```diff
@@ -97,7 +105,7 @@ def view_as_real(data):
 
 def fft2(
     data: torch.Tensor,
-    dim: Tuple[int, ...] = (1, 2),
+    dim: tuple[int, ...] = (1, 2),
     centered: bool = True,
```
Same

georgeyiasemis replied:
Done

jonasteuwen commented on docker/Dockerfile:
```diff
@@ -1,6 +1,6 @@
 ARG CUDA="11.3.0"
 ARG PYTORCH="1.10"
-ARG PYTHON="3.8"
+ARG PYTHON="3.9"
 
```
3.10

georgeyiasemis replied:
Will fix in future PR

georgeyiasemis replied:
Needs a  bit of work to get it right

jonasteuwen commented on docker/Dockerfile:
```diff
@@ -1,6 +1,6 @@
-ARG PYTHON="3.8"
+ARG PYTHON="3.9"
 
 # TODO: conda installs its own version of cuda
 FROM nvidia/cuda:${CUDA}-devel-ubuntu18.04
```
20.04

jonasteuwen commented on setup.py:
```diff
@@ -36,27 +36,26 @@ def finalize_options(self):
         "numpy>=1.21.2",
-        "h5py==3.3.0",
+        "h5py==3.11.0",
         "omegaconf==2.1.1",
         "torch>=1.10.2",
         "torchvision",
```
Still old version of torch, can you update? Also add explicit versions for torchvision. Same with OmegaConf

georgeyiasemis replied:
Done

jonasteuwen commented on tests/tests_data/test_mri_transforms.py:
```diff
@@ -262,6 +266,41 @@ def test_CropKspace(
+    ],
+)
+def test_RescaleKspace(shape, pad_shape, mode):
+    sample = create_sample(shape=shape + (2,))
+    transform = RescaleKspace(pad_shape, rescale_mode=mode)
+
+    sample = transform(sample)
+    assert sample["kspace"].shape == shape[: -len(pad_shape)] + pad_shape + (2,)
```
Don't you want to test if the interpolation is done well?

georgeyiasemis replied:
I am using pytorch's interpolation so I cannot really check that unless I compare to pytorch's interpolation, (if it makes sense what I am saying )

jonasteuwen reviewed (commented):
Some comments

georgeyiasemis commented on .github/workflows/pylint.yml:
```diff
@@ -7,10 +7,10 @@ jobs:
+    - name: Set up Python 3.9
       uses: actions/setup-python@v1
       with:
-        python-version: 3.8
+        python-version: '3.9'
     - name: Install dependencies
       run: |
```
Will address in seperate PR, as pylint for 3.10 throws some false positive errors.

georgeyiasemis replied:
Fixed for 3.10

georgeyiasemis commented on .github/workflows/tox.yml:
```diff
@@ -9,12 +9,12 @@ jobs:
     runs-on: ubuntu-latest
     strategy:
       matrix:
-        python-version: [3.8, 3.9]
+        python-version: ['3.9', '3.10']
 
     steps:
```
For py 3.11 some stuff fail in tests. I will do in seperate PR, but I think  moving from 3.8 and 3.9 to 3.9 and 3.10 suffices for now

georgeyiasemis commented on .github/workflows/coverage.yml:
```diff
@@ -8,7 +8,7 @@ jobs:
       matrix:
-        python-version: [3.8]
+        python-version: ['3.9']
     steps:
```
Done

georgeyiasemis commented:
@jonasteuwen This is ready for merge.
Some updates:
* I fixed pylint with py 9.10 (required a small import from skimage change)
* For some reason it threw an issue for cnp.int_t in cython which I replaced witg cnp.int64_t
* For docker, it needs more work and will do a seperate PR

jonasteuwen reviewed (approved):
LGTM. Nice work!
